AmazonDRS.agent.lib.nut

```diff
@@ -0,0 +1,416 @@
+    //     deviceModel : String         Device Model.
+    //     deviceSerial : String        Device Serial.
+    //     onAuthenticated : Function   Callback called when the operation is completed or an error happens.
+    //          (optional)              The callback signature:
+    //                                  onAuthenticated(error, response), where
+    //                                      error : Integer     0 if the authentication is successful, an error code otherwise.
+    //                                      response : Table    Key-value table with the response provided by Amazon server. May be null.
+    //     testDevice : Boolean         True if it is a test device. False by default.
+    //          (optional)
+    //
+    // Returns:                         Nothing.
+    function login(rocky, deviceModel, deviceSerial, onAuthenticated = null, testDevice = false) {
+        if (_loginEnabled) {
+            onAuthenticated && onAuthenticated(AMAZON_DRS_ERROR_LOGIN_ALREADY_CALLED, null);
+            return;
+        }
+
+        local authDone = function (error, resp) {
+            _loginEnabled = false;
+            _undefineLoginEndpoint(rocky);
+            onAuthenticated && onAuthenticated(error, resp);
+        }.bindenv(this);
+
+        _loginEnabled = true;
+        _defineLoginEndpoint(rocky, deviceModel, deviceSerial, testDevice, authDone);
+    }
+
+    // Allows to set a Refresh Token manually.
+    //
+    // Parameters:
+    //     refreshToken : String        A Refresh Token used to acquire an Access Token and refresh it when expired.
+    //
+    // Returns:                         Nothing.
+    function setRefreshToken(refreshToken) {
+        // Invalidate the existing Access Token
+        _accessToken = null;
+        _accessTokenExpiration = 0;
+        _refreshToken = refreshToken;
+    }
+
+    // Returns a string with the Refresh Token or null if it is not set.
+    //
+    // Returns:                         A string with the Refresh Token or null if it is not set.
+    function getRefreshToken() {
+        return _refreshToken;
+    }
+
+    // Places an order for a device/slot combination.
+    //
+    // Parameters:
+    //     slotId : String              ID of a slot to place an order for it.
+    //     onReplenished : Function     Callback called when the operation is completed or an error happens.
+    //          (optional)              The callback signature:
+    //                                  onReplenished(error, response), where
+    //                                      error : Integer     0 if the authentication is successful, an error code otherwise.
+    //                                      response : Table    Key-value table with the response provided by Amazon server. May be null.
+    //
+    // Returns:                         Nothing.
+    function replenish(slotId, onReplenished = null) {
+        if (_refreshToken == null) {
```

Comment:
replenish and cancelTestOrder have duplicated code (refreshToken and _isAccessTokenExpired conditionals)

Reply:
Fixed it